src/hotspot/share/gc/g1/g1CardSet.cpp

```diff
@@ -0,0 +1,899 @@
+    case CardSetBitMap: return card_set_ptr<G1CardSetBitMap>(card_set)->contains(card_in_region, _config->num_cards_in_howl_bitmap());
+    case CardSetHowl: {
+      G1CardSetHowl* howling_array = card_set_ptr<G1CardSetHowl>(card_set);
+
+      return howling_array->contains(card_in_region, _config);
+    }
+  }
+  ShouldNotReachHere();
+  return false;
+}
+
+void G1CardSet::print_info(outputStream* st, uint card_region, uint card_in_region) {
+  CardSetPtr card_set = get_card_set(card_region);
+  if (card_set == nullptr) {
+    st->print("NULL card set");
+    return;
+  } else if (card_set == FullCardSet) {
+    st->print("FULL card set)");
+    return;
+  }
+  switch (card_set_type(card_set)) {
+    case CardSetInlinePtr: {
+      st->print("InlinePtr not containing %u", card_in_region);
+      break;
+    }
+    case CardSetArrayOfCards :  {
+      st->print("AoC not containing %u", card_in_region);
+      break;
+    }
+    case CardSetBitMap: {
+      st->print("BitMap not containing %u", card_in_region);
+      break;
+    }
+    case CardSetHowl: {
+      st->print("CardSetHowl not containing %u", card_in_region);
+      break;
+    }
+    default: st->print("Unknown card set type %u", card_set_type(card_set)); ShouldNotReachHere(); break;
+  }
+}
+
+template <class CardVisitor>
+void G1CardSet::iterate_cards_during_transfer(CardSetPtr const card_set, CardVisitor& found) {
+  uint type = card_set_type(card_set);
+  assert(type == CardSetInlinePtr || type == CardSetArrayOfCards,
+         "invalid card set type %d to transfer from",
+         card_set_type(card_set));
+
+  switch (type) {
+    case CardSetInlinePtr: {
+      G1CardSetInlinePtr ptr(card_set);
+      ptr.iterate(found, _config->inline_ptr_bits_per_card());
+      return;
+    }
+    case CardSetArrayOfCards : {
+      card_set_ptr<G1CardSetArray>(card_set)->iterate(found);
+      return;
+    }
+    default:
+      ShouldNotReachHere();
+  }
+}
+
+void G1CardSet::iterate_containers(G1CardSetPtrIterator* found, bool at_safepoint) {
+  if (at_safepoint) {
+    _table->iterate_safepoint(found);
+  } else {
+    _table->iterate(found);
+  }
+}
+
+template <typename Closure>
+class G1ContainerCards {
+  Closure& _iter;
+  uint _region_idx;
+
+public:
+  G1ContainerCards(Closure& iter, uint region_idx) : _iter(iter), _region_idx(region_idx) { }
+
+  bool start_iterate(uint tag) { return true; }
+
+  void operator()(uint card_idx) {
+    _iter.do_card(_region_idx, card_idx);
+  }
+
+  void operator()(uint card_idx, uint length) {
+    for (uint i = 0; i < length; i++) {
+      _iter.do_card(_region_idx, card_idx);
+    }
+  }
+};
+
+void G1CardSet::iterate_cards(G1CardSetCardIterator& iter) {
+  G1CardSetMergeCardIterator<G1CardSetCardIterator, G1ContainerCards> cl(this, iter);
+  iterate_containers(&cl);
+}
+
+bool G1CardSet::occupancy_less_or_equal_to(size_t limit) const {
+  return occupied() <= limit;
+}
+
```

Comment:
Unused.
```suggestion
```

Reply:
I forgot to forward to this method in `HeapRegionRemSet::occupancy_less_or_equal_than()`.